v2/dtos/deviceprofile.go

```diff
@@ -54,3 +57,76 @@ func FromDeviceProfileModelToDTO(deviceProfile models.DeviceProfile) DeviceProfi
 		CoreCommands:    FromCommandModelsToDTOs(deviceProfile.CoreCommands),
 	}
 }
+
+func ValidateDeviceProfileDTO(profile DeviceProfile) error {
+	// deviceResources should not duplicated
+	for i := 0; i < len(profile.DeviceResources); i++ {
+		for j := i + 1; j < len(profile.DeviceResources); j++ {
+			if profile.DeviceResources[i].Name == profile.DeviceResources[j].Name {
+				return errors.NewCommonEdgeX(errors.KindContractInvalid, fmt.Sprintf("device resource %s is duplicated", profile.DeviceResources[j].Name), nil)
+			}
+		}
+	}
+	// deviceCommands should not duplicated
+	for i := 0; i < len(profile.DeviceCommands); i++ {
+		for j := i + 1; j < len(profile.DeviceCommands); j++ {
+			if profile.DeviceCommands[i].Name == profile.DeviceCommands[j].Name {
+				return errors.NewCommonEdgeX(errors.KindContractInvalid, fmt.Sprintf("device command %s is duplicated", profile.DeviceCommands[j].Name), nil)
+			}
+		}
+	}
+	// coreCommands should not duplicated
+	for i := 0; i < len(profile.CoreCommands); i++ {
+		for j := i + 1; j < len(profile.CoreCommands); j++ {
+			if profile.CoreCommands[i].Name == profile.CoreCommands[j].Name {
+				return errors.NewCommonEdgeX(errors.KindContractInvalid, fmt.Sprintf("core command %s is duplicated", profile.CoreCommands[j].Name), nil)
+			}
+		}
+	}
+	// coreCommands should match the one of deviceResources and deviceCommands
+	for i := 0; i < len(profile.CoreCommands); i++ {
+		match := false
+		for j := 0; j < len(profile.DeviceCommands); j++ {
+			if profile.CoreCommands[i].Name == profile.DeviceCommands[j].Name {
+				match = true
+				break
+			}
+		}
+		for j := 0; j < len(profile.DeviceResources); j++ {
+			if profile.CoreCommands[i].Name == profile.DeviceResources[j].Name {
+				match = true
+				break
+			}
+		}
+		if !match {
+			return errors.NewCommonEdgeX(errors.KindContractInvalid, fmt.Sprintf("core command %s doesn't match any deivce command or resource", profile.CoreCommands[i].Name), nil)
+		}
+	}
+	// deviceResources referenced in deviceCommands must exist
+	for i := 0; i < len(profile.DeviceCommands); i++ {
```

Comment:
Please use `for _, command := range profile.DeviceCommands`  instead of old style for loop.

Reply:
Refactored.